File: Java/Eta/Applications/Shared/src/main/java/com/refinitiv/eta/shared/rdm/marketbyprice/MarketByPriceRefresh.java
package com.refinitiv.eta.shared.rdm.marketbyprice;

import com.refinitiv.eta.codec.CodecFactory;
import com.refinitiv.eta.codec.CodecReturnCodes;
import com.refinitiv.eta.codec.DataDictionary;
import com.refinitiv.eta.codec.DataTypes;
import com.refinitiv.eta.codec.DictionaryEntry;
import com.refinitiv.eta.codec.EncodeIterator;
import com.refinitiv.eta.codec.Map;
import com.refinitiv.eta.codec.MapEntryActions;
import com.refinitiv.eta.codec.Msg;
import com.refinitiv.eta.codec.MsgClasses;
import com.refinitiv.eta.codec.Qos;
import com.refinitiv.eta.codec.RefreshMsg;
import com.refinitiv.eta.codec.State;
import com.refinitiv.eta.shared.rdm.marketbyorder.MarketByOrderItem;
import com.refinitiv.eta.shared.rdm.marketprice.MarketPriceRefreshFlags;
import com.refinitiv.eta.rdm.DomainTypes;
import com.refinitiv.eta.rdm.InstrumentNameTypes;

/**
 * The market by price refresh message. Used by an OMM Non-Interactive Provider
 * and OMM Interactive provider to encode/decode a market by order refresh
 * message.
 */
public class MarketByPriceRefresh extends MarketByPriceResponseBase
{
    private int serviceId;
    private int flags;
    private State state;
    private Qos qos;

    // used for encoding
    private RefreshMsg refreshMsg = (RefreshMsg)CodecFactory.createMsg();

    /**
     * Instantiates a new market by price refresh.
     */
    public MarketByPriceRefresh()
    {
        state = CodecFactory.createState();
        qos = CodecFactory.createQos();
    }

    /**
     * Clears the current contents of this object and prepares it for re-use.
     */
    public void clear()
    {
        super.clear();
        state.clear();
        qos.clear();
        flags = 0;
    }

    /**
     * The market by price refresh flags. Populated by
     * {@link MarketPriceRefreshFlags}.
     *
     * @return the int
     */
    public int flags()
    {
        return flags;
    }

    /**
     * The market by price refresh flags. Populated by
     * {@link MarketPriceRefreshFlags}.
     *
     * @param flags the flags
     */
    public void flags(int flags)
    {
        this.flags = flags;
    }

    /**
     * state of the market by order stream.
     * 
     * @return {@link State}
     */
    public State state()
    {
        return state;
    }

    /**
     * qos value of the market by order stream.
     * 
     * @return {@link Qos}
     */
    public Qos qos()
    {
        return qos;
    }

    /**
     * Service id.
     *
     * @return service id
     */
    public int serviceId()
    {
        return serviceId;
    }

    /**
     * Service id.
     *
     * @param serviceId the service id
     */
    public void serviceId(int serviceId)
    {
        this.serviceId = serviceId;
    }

    /* (non-Javadoc)
     * @see com.refinitiv.eta.shared.rdm.marketbyprice.MarketByPriceResponseBase#encodeMsg()
     */
    public Msg encodeMsg()
    {
        refreshMsg.clear();
        refreshMsg.msgClass(MsgClasses.REFRESH);
        refreshMsg.streamId(streamId());
        refreshMsg.domainType(DomainTypes.MARKET_BY_PRICE);
        refreshMsg.containerType(DataTypes.MAP);
        refreshMsg.applyHasMsgKey();

        if (checkRefreshComplete())
        {
            refreshMsg.applyRefreshComplete();
        }
        
        if (checkSolicited())
        {
            refreshMsg.applySolicited();
        }

        if (checkPrivateStream())
        {
            refreshMsg.applyPrivateStream();
        }
        
        if(checkClearCache())
        {
            refreshMsg.applyClearCache();
        }

        if (checkHasServiceId())
        {
            refreshMsg.msgKey().applyHasServiceId();
            refreshMsg.msgKey().serviceId(serviceId());
        }
        refreshMsg.msgKey().applyHasName();
        refreshMsg.msgKey().applyHasNameType();

        refreshMsg.state().dataState(state().dataState());
        refreshMsg.state().streamState(state().streamState());
        refreshMsg.state().code(state().code());
        refreshMsg.state().text(state().text());

        /* Itemname */
        refreshMsg.msgKey().name(itemName());
        refreshMsg.msgKey().nameType(InstrumentNameTypes.RIC);

        if (checkHasQos())
        {
            refreshMsg.applyHasQos();
            refreshMsg.qos().dynamic(qos().isDynamic());
            refreshMsg.qos().rate(qos().rate());
            refreshMsg.qos().timeliness(qos().timeliness());
            refreshMsg.qos().rateInfo(qos().rateInfo());
            refreshMsg.qos().timeInfo(qos().timeInfo());
        }

        return refreshMsg;
    }

    /**
     * Checks the presence of refresh complete flag.
     * 
     * This flag can also be bulk-get by {@link #flags()}
     * 
     * @return true - if exists; false if does not exist.
     */
    public boolean checkRefreshComplete()
    {
        return (flags & MarketPriceRefreshFlags.REFRESH_COMPLETE) != 0;
    }

    /**
     * Applies refresh complete flag.
     * 
     * This flag can also be bulk-set by {@link #flags(int)}
     */
    public void applyRefreshComplete()
    {
        flags |= MarketPriceRefreshFlags.REFRESH_COMPLETE;
    }

    /**
     * Checks the presence of solicited flag.
     * 
     * This flag can also be bulk-get by {@link #flags()}
     * 
     * @return true - if exists; false if does not exist.
     */
    public boolean checkSolicited()
    {
        return (flags & MarketPriceRefreshFlags.SOLICITED) != 0;
    }

    /**
     * Applies solicited flag.
     * 
     * This flag can also be bulk-set by {@link #flags(int)}
     */
    public void applySolicited()
    {
        flags |= MarketPriceRefreshFlags.SOLICITED;
    }

    /**
     * Checks the presence of clear cache flag.
     * 
     * This flag can also be bulk-get by {@link #flags()}
     * 
     * @return true - if exists; false if does not exist.
     */
    public boolean checkClearCache()
    {
        return (flags & MarketPriceRefreshFlags.CLEAR_CACHE) != 0;
    }

    /**
     * Applies clear cache flag.
     * 
     * This flag can also be bulk-set by {@link #flags(int)}
     */
    public void applyClearCache()
    {
        flags |= MarketPriceRefreshFlags.CLEAR_CACHE;
    }
    
    /**
     * Checks the presence of private stream flag.
     * 
     * This flag can also be bulk-get by {@link #flags()}
     * 
     * @return true - if exists; false if does not exist.
     */
    public boolean checkPrivateStream()
    {
        return (flags & MarketPriceRefreshFlags.PRIVATE_STREAM) != 0;
    }

    /**
     * Applies private stream flag.
     * 
     * This flag can also be bulk-set by {@link #flags(int)}
     */
    public void applyPrivateStream()
    {
        flags |= MarketPriceRefreshFlags.PRIVATE_STREAM;
    }

    /**
     * Checks the presence of Qos flag.
     * 
     * This flag can also be bulk-get by {@link #flags()}
     * 
     * @return true - if exists; false if does not exist.
     */
    public boolean checkHasQos()
    {
        return (flags & MarketPriceRefreshFlags.HAS_QOS) != 0;
    }

    /**
     * Applies qos flag.
     * 
     * This flag can also be bulk-set by {@link #flags(int)}
     */
    public void applyHasQos()
    {
        flags |= MarketPriceRefreshFlags.HAS_QOS;
    }

    /**
     * Checks the presence of service id flag.
     * 
     * This flag can also be bulk-get by {@link #flags()}
     * 
     * @return true - if exists; false if does not exist.
     */
    public boolean checkHasServiceId()
    {
        return (flags & MarketPriceRefreshFlags.HAS_SERVICE_ID) != 0;
    }

    /**
     * Applies service id flag.
     */
    public void applyHasServiceId()
    {
        flags |= MarketPriceRefreshFlags.HAS_SERVICE_ID;
    }

    /* (non-Javadoc)
     * @see com.refinitiv.eta.valueadd.domainrep.rdm.MsgBaseImpl#toString()
     */
    public String toString()
    {
        StringBuilder stringBuf = super.buildStringBuffer();
        stringBuf.insert(0, "MarketByPriceRefresh: \n");

        stringBuf.append(tabChar);
        stringBuf.append("state: ");
        stringBuf.append(state());
        stringBuf.append(eolChar);

        stringBuf.append(tabChar);
        stringBuf.append("solicited: ");
        stringBuf.append(checkSolicited());
        stringBuf.append(eolChar);

        stringBuf.append(tabChar);
        stringBuf.append("refreshComplete: ");
        stringBuf.append(checkRefreshComplete());
        stringBuf.append(eolChar);

        if (checkHasServiceId())
        {
            stringBuf.append(tabChar);
            stringBuf.append("serviceId: ");
            stringBuf.append(serviceId());
            stringBuf.append(eolChar);
        }
        if (checkHasQos())
        {
            stringBuf.append(tabChar);
            stringBuf.append("qos: ");
            stringBuf.append(refreshMsg.qos().toString());
            stringBuf.append(eolChar);
        }

        return stringBuf.toString();
    }

    /* (non-Javadoc)
     * @see com.refinitiv.eta.shared.rdm.marketbyprice.MarketByPriceResponseBase#encodeSummaryData(com.refinitiv.eta.codec.EncodeIterator, com.refinitiv.eta.codec.DataDictionary, com.refinitiv.eta.codec.Map)
     */
    protected int encodeSummaryData(EncodeIterator encodeIter, DataDictionary dictionary, Map map)
    {
        DictionaryEntry dictionaryEntry;
        int ret = CodecReturnCodes.SUCCESS;

        /* encode fields for summary data */
        fieldList.clear();
        fieldList.applyHasStandardData();
        if ((ret = fieldList.encodeInit(encodeIter, null, 0)) < CodecReturnCodes.SUCCESS)
        {
            return ret;
        }

        /* CURRENCY */
        fieldEntry.clear();
        dictionaryEntry = dictionary.entry(MarketByPriceItem.CURRENCY_FID);
        if (dictionaryEntry != null)
        {
            fieldEntry.fieldId(MarketByOrderItem.CURRENCY_FID);
            fieldEntry.dataType(dictionaryEntry.rwfType());
            if ((ret = fieldEntry.encode(encodeIter, MarketByOrderItem.USD_ENUM)) < CodecReturnCodes.SUCCESS)
            {
                return ret;
            }
        }

        /* MARKET STATUS INDICATOR */
        fieldEntry.clear();
        dictionaryEntry = dictionary.entry(MarketByOrderItem.MKT_STATUS_IND_FID);
        if (dictionaryEntry != null)
        {
            fieldEntry.fieldId(MarketByOrderItem.MKT_STATUS_IND_FID);
            fieldEntry.dataType(dictionaryEntry.rwfType());
            if ((ret = fieldEntry.encode(encodeIter, MarketByOrderItem.BBO_ENUM)) < CodecReturnCodes.SUCCESS)
            {
                return ret;
            }
        }

        /* ACTIVE DATE */
        fieldEntry.clear();
        dictionaryEntry =
                dictionary.entry(MarketByOrderItem.ACTIVE_DATE_FID);
        if (dictionaryEntry != null)
        {
            fieldEntry.fieldId(MarketByOrderItem.ACTIVE_DATE_FID);
            fieldEntry.dataType(dictionaryEntry.rwfType());
            tempDateTime.localTime();
            if ((ret = fieldEntry.encode(encodeIter, tempDateTime.date())) < CodecReturnCodes.SUCCESS)
            {
                return ret;
            }
        }

        ret = fieldList.encodeComplete(encodeIter, true);
        return ret;
    }

    /* (non-Javadoc)
     * @see com.refinitiv.eta.shared.rdm.marketbyprice.MarketByPriceResponseBase#encodeMapEntries(com.refinitiv.eta.codec.EncodeIterator, com.refinitiv.eta.codec.DataDictionary)
     */
    protected int encodeMapEntries(EncodeIterator encodeIter, DataDictionary dictionary)
    {
        // encode the order book
        MarketByPriceItem.PriceInfo priceInfo = marketByPriceItem().priceInfoList.get(partNo());
        mapEntry.clear();
        mapEntry.action(MapEntryActions.ADD);
        tmpBuffer.data(priceInfo.PRICE_POINT);
        int ret;
        if ((ret = mapEntry.encodeInit(encodeIter, tmpBuffer, 0)) != CodecReturnCodes.SUCCESS)
        {
            return ret;
        }

        fieldList.clear();
        fieldList.applyHasStandardData();
        fieldList.applyHasSetData();
        fieldList.applyHasSetId();
        fieldList.setId(0);

        ret = fieldList.encodeInit(encodeIter, marketByPriceSetDefDb, 0);
        if (ret < CodecReturnCodes.SUCCESS)
        {
            return ret;
        }

        // encode fields

        // ORDER_PRC
        fieldEntry.clear();
        DictionaryEntry dictionaryEntry = dictionary.entry(MarketByPriceItem.ORDER_PRC_FID);
        if (dictionaryEntry != null)
        {
            fieldEntry.fieldId(MarketByPriceItem.ORDER_PRC_FID);
            fieldEntry.dataType(dictionaryEntry.rwfType());
            ret = fieldEntry.encode(encodeIter, priceInfo.ORDER_PRC);
            if (ret < CodecReturnCodes.SUCCESS)
            {
                return ret;
            }
        }

        // ORDER_SIZE
        fieldEntry.clear();
        dictionaryEntry = dictionary.entry(MarketByPriceItem.ORDER_SIZE_FID);
        if (dictionaryEntry != null)
        {
            fieldEntry.fieldId(MarketByPriceItem.ORDER_SIZE_FID);
            fieldEntry.dataType(dictionaryEntry.rwfType());
            ret = fieldEntry.encode(encodeIter, priceInfo.ORDER_SIZE);
            if (ret < CodecReturnCodes.SUCCESS)
            {
                return ret;
            }
        }

        // QUOTIM_MS
        fieldEntry.clear();
        dictionaryEntry = dictionary.entry(MarketByPriceItem.QUOTIM_MS_FID);
        if (dictionaryEntry != null)
        {
            fieldEntry.fieldId(MarketByPriceItem.QUOTIM_MS_FID);
            fieldEntry.dataType(dictionaryEntry.rwfType());
            tempUInt.value(priceInfo.QUOTIM_MS);
            ret = fieldEntry.encode(encodeIter, tempUInt);
            if (ret < CodecReturnCodes.SUCCESS)
            {
                return ret;
            }
        }

        // NO_ORD
        fieldEntry.clear();
        dictionaryEntry = dictionary.entry(MarketByPriceItem.NO_ORD_FID);
        if (dictionaryEntry != null)
        {
            fieldEntry.fieldId(MarketByPriceItem.NO_ORD_FID);
            fieldEntry.dataType(dictionaryEntry.rwfType());
            tempUInt.value(priceInfo.NO_ORD);
            /*
             * This encoding completes the encoding of the ORDER_PRC,
             * ORDER_SIZE, QUOTIM_MS, NO_ORD set. When a set is completed, a
             * success code, CodecReturnCodes.SET_COMPLETE, is returned to
             * indicate this. This may be used to ensure that the set is being
             * used as intended.
             */
            ret = fieldEntry.encode(encodeIter, tempUInt);
            if (ret != CodecReturnCodes.SET_COMPLETE)
            {
                return CodecReturnCodes.FAILURE;
            }
        }

        // encode items for private stream
        if (checkPrivateStream())
        {
            // MKOASK_VOL
            fieldEntry.clear();
            dictionaryEntry = dictionary.entry(MarketByPriceItem.MKOASK_VOL_FID);
            if (dictionaryEntry != null)
            {
                fieldEntry.fieldId(MarketByPriceItem.MKOASK_VOL_FID);
                fieldEntry.dataType(dictionaryEntry.rwfType());
                ret = fieldEntry.encode(encodeIter, priceInfo.MKOASK_VOL);
                if (ret < CodecReturnCodes.SUCCESS)
                {
                    return ret;
                }
            }

            // MKOBID_VOL
            fieldEntry.clear();
            dictionaryEntry = dictionary.entry(MarketByPriceItem.MKOBID_VOL_FID);
            if (dictionaryEntry != null)
            {
                fieldEntry.fieldId(MarketByPriceItem.MKOBID_VOL_FID);
                fieldEntry.dataType(dictionaryEntry.rwfType());
                ret = fieldEntry.encode(encodeIter, priceInfo.MKOBID_VOL);
                if (ret < CodecReturnCodes.SUCCESS)
                {
                    return ret;
                }
            }
        }
        /*
         * Since the field list specified FieldListFlags.HAS_STANDARD_DATA as
         * well, standard entries can be encoded after the set is finished. The
         * Add actions will additionally include the ORDER_SIDE field.
         */
        // ORDER_SIDE
        fieldEntry.clear();
        dictionaryEntry = dictionary.entry(MarketByPriceItem.ORDER_SIDE_FID);
        if (dictionaryEntry != null)
        {
            fieldEntry.fieldId(MarketByPriceItem.ORDER_SIDE_FID);
            fieldEntry.dataType(dictionaryEntry.rwfType());
            ret = fieldEntry.encode(encodeIter, priceInfo.ORDER_SIDE);
            if (ret < CodecReturnCodes.SUCCESS)
            {
                return ret;
            }
        }

        ret = fieldList.encodeComplete(encodeIter, true);
        if (ret != CodecReturnCodes.SUCCESS)
        {
            return ret;
        }

        return mapEntry.encodeComplete(encodeIter, true);
    }
    
    /* (non-Javadoc)
     * @see com.refinitiv.eta.valueadd.domainrep.rdm.MsgBase#domainType()
     */
    @Override
    public int domainType()
    {
        return DomainTypes.MARKET_BY_PRICE;
    }
}

File: CSharp/Eta/Applications/EtaAppCommon/Rdm/item/marketbyprice/MarketByPriceRefresh.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license      --
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
 *|                See the project's LICENSE.md for details.                  --
 *|           Copyright (C) 2022 Refinitiv. All rights reserved.              --
 *|-----------------------------------------------------------------------------
 */

using Refinitiv.Eta.Codec;
using Refinitiv.Eta.Rdm;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;

namespace Refinitiv.Eta.Example.Common
{
    public class MarketByPriceRefresh : MarketByPriceBase
    {
        private Qos qos = new Qos();
        public int ServiceId { get; set; }

        public override int StreamId { get => m_RefreshMsg.StreamId; set { m_RefreshMsg.StreamId = value; } }
        public override int MsgClass { get => m_RefreshMsg.MsgClass; }
        public override int DomainType { get => m_RefreshMsg.DomainType; }

        public ItemDomainCommonFlags Flags { get; set; }

        public bool Solicited
        {
            get => (Flags & ItemDomainCommonFlags.SOLICITED) != 0;
            set
            {
                if (value)
                {
                    Flags |= ItemDomainCommonFlags.SOLICITED;
                    m_RefreshMsg.ApplySolicited();
                }
                else
                    Flags &= ~ItemDomainCommonFlags.SOLICITED;
            }
        }
        public bool RefreshComplete
        {
            get => (Flags & ItemDomainCommonFlags.REFRESH_COMPLETE) != 0;
            set
            {
                if (value)
                {
                    Flags |= ItemDomainCommonFlags.REFRESH_COMPLETE;
                }
                else
                    Flags &= ~ItemDomainCommonFlags.REFRESH_COMPLETE;
            }
        }
        public bool ClearCache
        {
            get => (Flags & ItemDomainCommonFlags.CLEAR_CACHE) != 0;
            set
            {
                if (value) { Flags |= ItemDomainCommonFlags.CLEAR_CACHE; }
                else Flags &= ~ItemDomainCommonFlags.CLEAR_CACHE;
            }
        }
        public bool HasQos
        {
            get => (Flags & ItemDomainCommonFlags.HAS_QOS) != 0;
            set
            {
                if (value)
                {
                    Flags |= ItemDomainCommonFlags.HAS_QOS;
                }
                else
                    Flags &= ~ItemDomainCommonFlags.HAS_QOS;
            }
        }
        public bool HasServiceId
        {
            get => (Flags & ItemDomainCommonFlags.HAS_SERVICE_ID) != 0;
            set
            {
                if (value)
                {
                    Flags |= ItemDomainCommonFlags.HAS_SERVICE_ID;
                }
                else
                    Flags &= ~ItemDomainCommonFlags.HAS_SERVICE_ID;
            }
        }
        public bool PrivateStream
        {
            get => (Flags & ItemDomainCommonFlags.PRIVATE_STREAM) != 0;
            set
            {
                if (value)
                {
                    Flags |= ItemDomainCommonFlags.PRIVATE_STREAM;
                }
                else
                    Flags &= ~ItemDomainCommonFlags.SOLICITED;
            }
        }
        public State State
        {
            get => m_RefreshMsg.State;
            set
            {
                Debug.Assert(value != null);
                value.Copy(m_RefreshMsg.State);
            }
        }
        public Qos Qos
        {
            get => qos;
            set
            {
                Debug.Assert(value != null);
                value.Copy(qos);
            }
        }
        public long Filter { get => m_RefreshMsg.MsgKey.Filter; set { m_RefreshMsg.MsgKey.Filter = value; } }

        private IRefreshMsg m_RefreshMsg = new Msg();

        public MarketByPriceRefresh()
        {
            Clear();
        }

        public override Msg EncodeMsg()
        {
            if (RefreshComplete)
            {
                m_RefreshMsg.ApplyRefreshComplete();
            }

            if (Solicited)
            {
                m_RefreshMsg.ApplySolicited();
            }

            if (PrivateStream)
            {
                m_RefreshMsg.ApplyPrivateStream();
            }

            if (ClearCache)
            {
                m_RefreshMsg.ApplyClearCache();
            }

            if (HasServiceId)
            {
                m_RefreshMsg.MsgKey.ApplyHasServiceId();
                m_RefreshMsg.MsgKey.ServiceId = ServiceId;
            }
            m_RefreshMsg.MsgKey.ApplyHasName();
            m_RefreshMsg.MsgKey.ApplyHasNameType();

            /* Itemname */
            m_RefreshMsg.MsgKey.Name = ItemName;
            m_RefreshMsg.MsgKey.NameType = InstrumentNameTypes.RIC;

            if (HasQos)
            {
                m_RefreshMsg.ApplyHasQos();
                qos.Copy(m_RefreshMsg.Qos);
            }

            return (Msg)m_RefreshMsg;
        }

        protected override CodecReturnCode EncodeMapEntries(EncodeIterator EncodeIter, DataDictionary dictionary)
        {
            // Encode the order book
            PriceInfo priceInfo = MbpInfo.PriceInfoList[PartNo];
            mapEntry.Clear();
            mapEntry.Action = MapEntryActions.ADD;
            tmpBuffer.Data(priceInfo.PRICE_POINT);
            CodecReturnCode ret;
            if ((ret = mapEntry.EncodeInit(EncodeIter, tmpBuffer, 0)) != CodecReturnCode.SUCCESS)
            {
                return ret;
            }

            fieldList.Clear();
            fieldList.ApplyHasStandardData();
            fieldList.ApplyHasSetData();
            fieldList.ApplyHasSetId();
            fieldList.SetId = 0;

            ret = fieldList.EncodeInit(EncodeIter, marketByPriceSetDefDb, 0);
            if (ret < CodecReturnCode.SUCCESS)
            {
                return ret;
            }

            // Encode fields

            // ORDER_PRC
            fieldEntry.Clear();
            IDictionaryEntry dictionaryEntry = dictionary.Entry(MarketByPriceItem.ORDER_PRC_FID);
            if (dictionaryEntry != null)
            {
                fieldEntry.FieldId = MarketByPriceItem.ORDER_PRC_FID;
                fieldEntry.DataType = dictionaryEntry.RwfType;
                ret = fieldEntry.Encode(EncodeIter, priceInfo.ORDER_PRC);
                if (ret < CodecReturnCode.SUCCESS)
                {
                    return ret;
                }
            }

            // ORDER_SIZE
            fieldEntry.Clear();
            dictionaryEntry = dictionary.Entry(MarketByPriceItem.ORDER_SIZE_FID);
            if (dictionaryEntry != null)
            {
                fieldEntry.FieldId = MarketByPriceItem.ORDER_SIZE_FID;
                fieldEntry.DataType = dictionaryEntry.RwfType;
                ret = fieldEntry.Encode(EncodeIter, priceInfo.ORDER_SIZE);
                if (ret < CodecReturnCode.SUCCESS)
                {
                    return ret;
                }
            }

            // QUOTIM_MS
            fieldEntry.Clear();
            dictionaryEntry = dictionary.Entry(MarketByPriceItem.QUOTIM_MS_FID);
            if (dictionaryEntry != null)
            {
                fieldEntry.FieldId = MarketByPriceItem.QUOTIM_MS_FID;
                fieldEntry.DataType = dictionaryEntry.RwfType;
                tempUInt.Value(priceInfo.QUOTIM_MS);
                ret = fieldEntry.Encode(EncodeIter, tempUInt);
                if (ret < CodecReturnCode.SUCCESS)
                {
                    return ret;
                }
            }

            // NO_ORD
            fieldEntry.Clear();
            dictionaryEntry = dictionary.Entry(MarketByPriceItem.NO_ORD_FID);
            if (dictionaryEntry != null)
            {
                fieldEntry.FieldId = MarketByPriceItem.NO_ORD_FID;
                fieldEntry.DataType = dictionaryEntry.RwfType;
                tempUInt.Value(priceInfo.NO_ORD);

                // This encoding completes the encoding of the ORDER_PRC,
                // ORDER_SIZE, QUOTIM_MS, NO_ORD set.
                ret = fieldEntry.Encode(EncodeIter, tempUInt);
                if (ret != CodecReturnCode.SUCCESS)
                {
                    return CodecReturnCode.FAILURE;
                }
            }

            // Encode items for private stream
            if (PrivateStream)
            {
                // MKOASK_VOL
                fieldEntry.Clear();
                dictionaryEntry = dictionary.Entry(MarketByPriceItem.MKOASK_VOL_FID);
                if (dictionaryEntry != null)
                {
                    fieldEntry.FieldId = MarketByPriceItem.MKOASK_VOL_FID;
                    fieldEntry.DataType = dictionaryEntry.RwfType;
                    ret = fieldEntry.Encode(EncodeIter, priceInfo.MKOASK_VOL);
                    if (ret < CodecReturnCode.SUCCESS)
                    {
                        return ret;
                    }
                }

                // MKOBID_VOL
                fieldEntry.Clear();
                dictionaryEntry = dictionary.Entry(MarketByPriceItem.MKOBID_VOL_FID);
                if (dictionaryEntry != null)
                {
                    fieldEntry.FieldId = MarketByPriceItem.MKOBID_VOL_FID;
                    fieldEntry.DataType = dictionaryEntry.RwfType;
                    ret = fieldEntry.Encode(EncodeIter, priceInfo.MKOBID_VOL);
                    if (ret < CodecReturnCode.SUCCESS)
                    {
                        return ret;
                    }
                }
            }
            /*
             * Since the field list specified FieldListFlags.HAS_STANDARD_DATA as
             * well, standard entries can be Encoded after the set is finished. The
             * Add actions will additionally include the ORDER_SIDE field.
             */
            // ORDER_SIDE
            fieldEntry.Clear();
            dictionaryEntry = dictionary.Entry(MarketByPriceItem.ORDER_SIDE_FID);
            if (dictionaryEntry != null)
            {
                fieldEntry.FieldId = MarketByPriceItem.ORDER_SIDE_FID;
                fieldEntry.DataType = dictionaryEntry.RwfType;
                ret = fieldEntry.Encode(EncodeIter, priceInfo.ORDER_SIDE);
                if (ret < CodecReturnCode.SUCCESS)
                {
                    return ret;
                }
            }

            ret = fieldList.EncodeComplete(EncodeIter, true);
            if (ret != CodecReturnCode.SUCCESS)
            {
                return ret;
            }

            return mapEntry.EncodeComplete(EncodeIter, true);
        }
        protected override CodecReturnCode EncodeSummaryData(EncodeIterator EncodeIter, DataDictionary dictionary, Map map)
        {
            IDictionaryEntry dictionaryEntry;
            CodecReturnCode ret = CodecReturnCode.SUCCESS;

            /* Encode fields for summary data */
            fieldList.Clear();
            fieldList.ApplyHasStandardData();
            if ((ret = fieldList.EncodeInit(EncodeIter, null, 0)) < CodecReturnCode.SUCCESS)
            {
                return ret;
            }

            /* CURRENCY */
            fieldEntry.Clear();
            dictionaryEntry = dictionary.Entry(MarketByPriceItem.CURRENCY_FID);
            if (dictionaryEntry != null)
            {
                fieldEntry.FieldId = MarketByOrderItem.CURRENCY_FID;
                fieldEntry.DataType = dictionaryEntry.RwfType;
                if ((ret = fieldEntry.Encode(EncodeIter, MarketByOrderItem.USD_ENUM)) < CodecReturnCode.SUCCESS)
                {
                    return ret;
                }
            }

            /* MARKET STATUS INDICATOR */
            fieldEntry.Clear();
            dictionaryEntry = dictionary.Entry(MarketByOrderItem.MKT_STATUS_IND_FID);
            if (dictionaryEntry != null)
            {
                fieldEntry.FieldId = MarketByOrderItem.MKT_STATUS_IND_FID;
                fieldEntry.DataType = dictionaryEntry.RwfType;
                if ((ret = fieldEntry.Encode(EncodeIter, MarketByOrderItem.BBO_ENUM)) < CodecReturnCode.SUCCESS)
                {
                    return ret;
                }
            }

            /* ACTIVE DATE */
            fieldEntry.Clear();
            dictionaryEntry = dictionary.Entry(MarketByOrderItem.ACTIVE_DATE_FID);
            if (dictionaryEntry != null)
            {
                fieldEntry.FieldId = MarketByOrderItem.ACTIVE_DATE_FID;
                fieldEntry.DataType = dictionaryEntry.RwfType;
                tempDateTime.LocalTime();
                if ((ret = fieldEntry.Encode(EncodeIter, tempDateTime.Date())) < CodecReturnCode.SUCCESS)
                {
                    return ret;
                }
            }

            ret = fieldList.EncodeComplete(EncodeIter, true);
            return ret;
        }

        public override void Clear()
        {
            base.Clear();
            m_RefreshMsg.Clear();
            m_RefreshMsg.MsgClass = MsgClasses.REFRESH;
            m_RefreshMsg.DomainType = (int)Rdm.DomainType.MARKET_BY_PRICE;
            m_RefreshMsg.ContainerType = DataTypes.MAP;
            m_RefreshMsg.ApplyHasMsgKey();
            qos.Clear();
        }

        public override string ToString()
        {
            StringBuilder stringBuf = PrepareStringBuilder();
            stringBuf.Insert(0, "MarketByPriceRefresh: \n");

            stringBuf.Append(tabChar);
            stringBuf.Append("state: ");
            stringBuf.Append(State);
            stringBuf.Append(eolChar);

            stringBuf.Append(tabChar);
            stringBuf.Append("solicited: ");
            stringBuf.Append(Solicited);
            stringBuf.Append(eolChar);

            stringBuf.Append(tabChar);
            stringBuf.Append("refreshComplete: ");
            stringBuf.Append(RefreshComplete);
            stringBuf.Append(eolChar);

            if (HasServiceId)
            {
                stringBuf.Append(tabChar);
                stringBuf.Append("serviceId: ");
                stringBuf.Append(ServiceId);
                stringBuf.Append(eolChar);
            }
            if (HasQos)
            {
                stringBuf.Append(tabChar);
                stringBuf.Append("qos: ");
                stringBuf.Append(qos.ToString());
                stringBuf.Append(eolChar);
            }

            return stringBuf.ToString();
        }
    }
}
